File: backends/cloudflare.py
"""Cloudflare speed test backend.

Uses the public speed.cloudflare.com endpoints (no auth). Global anycast, so
the test automatically hits the nearest Cloudflare PoP.

Connection info uses a graceful fallback chain because /meta started rejecting
plain Python user agents with HTTP 403:
  1. /meta with a real browser UA — full info: IP, ISP, city, colo
  2. /cdn-cgi/trace → IP, colo, ISO-2 country; enriched via ipwho.is for ISP+city
  3. /cdn-cgi/trace alone — IP, colo, country (no ISP)
  4. api.ipify.org — IP only

The "country" field in the fallback path is an ISO-2 code (e.g. "BD") whereas
/meta gives the two-letter code too; both are consistent ISO-2 for this backend.
"""
from __future__ import annotations

import json
import os
import urllib.parse
import urllib.request

from .base import (
    BROWSER_UA,
    NETWORK_EXCEPTIONS,
    Chunk,
    ConnectionInfo,
    LatencyResult,
    ProgressCallback,
    SpeedResult,
    SpeedTestBackend,
    http_get,
    measure_latency,
    run_chunks,
    safe_decode,
)

DOWN_URL = "https://speed.cloudflare.com/__down?bytes={}"
UP_URL = "https://speed.cloudflare.com/__up"
META_URL = "https://speed.cloudflare.com/meta"
TRACE_URL = "https://www.cloudflare.com/cdn-cgi/trace"
IPIFY_URL = "https://api.ipify.org?format=json"
IPWHO_URL = "https://ipwho.is/"

DOWN_SIZES = [
    (1_000_000, "1 MB"),
    (5_000_000, "5 MB"),
    (10_000_000, "10 MB"),
    (25_000_000, "25 MB"),
]
# Upload sizes symmetric with DOWN_SIZES so TCP has time to ramp up and
# saturate before the sample ends. With the old 500K-5M range, the top
# samples often measured warmup slope rather than steady-state throughput
# and reported ~30% low on fast links.
UP_SIZES = [
    (1_000_000, "1 MB"),
    (5_000_000, "5 MB"),
    (10_000_000, "10 MB"),
    (25_000_000, "25 MB"),
]


def _parse_trace(body: str) -> dict:
    out: dict = {}
    for line in body.splitlines():
        if "=" in line:
            k, v = line.split("=", 1)
            out[k.strip()] = v.strip()
    return out


def _ipwho_lookup(ip: str = "") -> dict | None:
    """Look up ISP/city via ipwho.is. Returns None on any failure.
    Empty IP lets the service auto-detect our public IP."""
    # quote() defends against malformed IPs from upstream — e.g. an IPv6
    # with a zone suffix or junk that sneaks past trace parsing. The empty
    # string passes through unchanged, preserving auto-detect behaviour.
    safe_ip = urllib.parse.quote(ip, safe=":")
    try:
        data = json.loads(safe_decode(http_get(IPWHO_URL + safe_ip, timeout=5)))
        if not data.get("success", True):
            return None
        connection = data.get("connection", {}) or {}
        return {
            "ip": data.get("ip", ""),
            "isp": connection.get("isp") or connection.get("org") or "Unknown",
            "city": data.get("city", ""),
            "region": data.get("region", ""),
            "country_code": data.get("country_code", ""),
        }
    except (*NETWORK_EXCEPTIONS, json.JSONDecodeError):
        return None


class CloudflareBackend(SpeedTestBackend):
    name = "cloudflare"
    display_name = "Cloudflare"

    def connection_info(self) -> ConnectionInfo:
        try:
            data = json.loads(safe_decode(http_get(META_URL, timeout=5)))
            return ConnectionInfo(
                ip=data.get("clientIp", "Unknown"),
                isp=data.get("asOrganization", "Unknown"),
                city=data.get("city", ""),
                region=data.get("region", "Unknown"),
                country=data.get("country", ""),
                server=f"Cloudflare {data.get('colo', '')}".strip(),
            )
        except (*NETWORK_EXCEPTIONS, json.JSONDecodeError):
            pass

        trace: dict = {}
        try:
            trace = _parse_trace(safe_decode(http_get(TRACE_URL, timeout=5)))
        except NETWORK_EXCEPTIONS:
            pass

        ip = trace.get("ip", "")
        colo = trace.get("colo", "")
        country = trace.get("loc", "")

        enriched = _ipwho_lookup(ip)
        if enriched:
            return ConnectionInfo(
                ip=ip or enriched.get("ip", "Unknown"),
                isp=enriched.get("isp", "Unknown"),
                city=enriched.get("city", ""),
                region=enriched.get("region", "Unknown"),
                country=country or enriched.get("country_code", ""),
                server=f"Cloudflare {colo}".strip() if colo else "Cloudflare",
            )

        if ip:
            return ConnectionInfo(
                ip=ip,
                country=country,
                server=f"Cloudflare {colo}".strip() if colo else "Cloudflare",
            )

        try:
            data = json.loads(safe_decode(http_get(IPIFY_URL, timeout=5)))
            return ConnectionInfo(ip=data.get("ip", "Unknown"), server="Cloudflare")
        except (*NETWORK_EXCEPTIONS, json.JSONDecodeError):
            return ConnectionInfo(server="Cloudflare")

    def test_latency(
        self, samples: int = 10, callback: ProgressCallback = None
    ) -> LatencyResult:
        def factory() -> urllib.request.Request:
            return urllib.request.Request(
                DOWN_URL.format(0), headers={"User-Agent": BROWSER_UA}
            )

        return measure_latency(factory, samples, callback, backend=self)

    def test_download(self, callback: ProgressCallback = None) -> SpeedResult:
        def make(size: int) -> urllib.request.Request:
            return urllib.request.Request(
                DOWN_URL.format(size), headers={"User-Agent": BROWSER_UA}
            )

        chunks = [
            Chunk(request_factory=lambda s=s: make(s), label=label)
            for s, label in DOWN_SIZES
        ]
        return run_chunks(chunks, "download_chunk", callback, backend=self)

    def test_upload(self, callback: ProgressCallback = None) -> SpeedResult:
        def make(size: int) -> urllib.request.Request:
            return urllib.request.Request(
                UP_URL,
                data=os.urandom(size),
                headers={
                    "User-Agent": BROWSER_UA,
                    "Content-Type": "application/octet-stream",
                },
                method="POST",
            )

        chunks = [
            Chunk(request_factory=lambda s=s: make(s), label=label, size_bytes=s)
            for s, label in UP_SIZES
        ]
        return run_chunks(chunks, "upload_chunk", callback, backend=self)

File: mobile/lib/backends/cloudflare.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:http/http.dart' as http;
import 'package:logging/logging.dart';

import '../models/connection_info.dart';
import '../models/latency_result.dart';
import '../models/speed_result.dart';
import '../utils/browser_ua.dart';
import 'backend_error.dart';
import 'progress_event.dart';
import 'speed_test_backend.dart';

final _log = Logger('Cloudflare');

const Duration _shortTimeout = Duration(seconds: 5);
const Duration _chunkTimeout = Duration(seconds: 30);
const int _progressEvery = 256 * 1024;
const Map<String, String> _browserHeaders = {'User-Agent': browserUa};

// More complete browser-like headers for /meta — helps bypass Cloudflare's
// bot-score rejection when plain UA alone isn't enough.
const Map<String, String> _metaHeaders = {
  'User-Agent': browserUa,
  'Accept': 'application/json,text/plain,*/*',
  'Accept-Language': 'en-US,en;q=0.9',
  'Sec-Fetch-Mode': 'cors',
  'Sec-Fetch-Site': 'same-origin',
  'Referer': 'https://speed.cloudflare.com/',
};

Map<String, String> _parseTrace(String body) {
  final out = <String, String>{};
  for (final line in const LineSplitter().convert(body)) {
    final eq = line.indexOf('=');
    if (eq > 0) out[line.substring(0, eq).trim()] = line.substring(eq + 1).trim();
  }
  return out;
}

/// Look up ISP + city via ipwho.is. Returns null on any failure.
/// Empty IP lets the service auto-detect our public IP.
Future<Map<String, String>?> _ipwhoLookup(String ip) async {
  try {
    final safeIp = Uri.encodeComponent(ip);
    final resp = await http
        .get(Uri.parse('$_ipwhoUrl$safeIp'), headers: _browserHeaders)
        .timeout(_shortTimeout);
    if (resp.statusCode != 200) return null;
    final data = jsonDecode(resp.body) as Map<String, dynamic>;
    if (data['success'] == false) return null;
    final connection = (data['connection'] as Map<String, dynamic>?) ?? const {};
    return {
      'ip': (data['ip'] as String?) ?? '',
      'isp': (connection['isp'] as String?) ??
          (connection['org'] as String?) ??
          'Unknown',
      'city': (data['city'] as String?) ?? '',
      'region': (data['region'] as String?) ?? '',
      'country_code': (data['country_code'] as String?) ?? '',
    };
  } catch (_) {
    return null;
  }
}

const String _downUrl = 'https://speed.cloudflare.com/__down?bytes={}';
const String _upUrl = 'https://speed.cloudflare.com/__up';
const String _metaUrl = 'https://speed.cloudflare.com/meta';
const String _traceUrl = 'https://www.cloudflare.com/cdn-cgi/trace';
const String _ipifyUrl = 'https://api.ipify.org?format=json';
const String _ipwhoUrl = 'https://ipwho.is/';

// (chunkBytes, label) — symmetric for download + upload so TCP has time to
// saturate. Asymmetric sizes under-measured fast links on the Linux parent.
const List<(int, String)> _downSizes = [
  (1000000, '1 MB'),
  (5000000, '5 MB'),
  (10000000, '10 MB'),
  (25000000, '25 MB'),
];
const List<(int, String)> _upSizes = _downSizes;

class CloudflareBackend extends SpeedTestBackend {
  @override
  String get name => 'cloudflare';

  @override
  String get displayName => 'Cloudflare';

  bool _cancelled = false;
  http.Client? _client;

  @override
  Future<ConnectionInfo> connectionInfo() async {
    try {
      final resp = await http
          .get(Uri.parse(_metaUrl), headers: _metaHeaders)
          .timeout(_shortTimeout);
      _log.info('/meta status=${resp.statusCode}');
      if (resp.statusCode == 200) {
        final data = jsonDecode(resp.body) as Map<String, dynamic>;
        final colo = (data['colo'] as String?) ?? '';
        return ConnectionInfo(
          ip: (data['clientIp'] as String?) ?? 'Unknown',
          isp: (data['asOrganization'] as String?) ?? 'Unknown',
          city: (data['city'] as String?) ?? '',
          region: (data['region'] as String?) ?? 'Unknown',
          country: (data['country'] as String?) ?? '',
          server: 'Cloudflare $colo'.trim(),
        );
      }
    } catch (e) {
      _log.info('/meta failed: $e — falling through to trace');
    }

    Map<String, String> trace = {};
    try {
      final resp = await http
          .get(Uri.parse(_traceUrl), headers: _browserHeaders)
          .timeout(_shortTimeout);
      if (resp.statusCode == 200) trace = _parseTrace(resp.body);
    } catch (_) {}

    final ip = trace['ip'] ?? '';
    final colo = trace['colo'] ?? '';
    final country = trace['loc'] ?? '';

    // Enrichment: ipwho.is gives us ISP + city that trace doesn't have.
    if (ip.isNotEmpty) {
      final enriched = await _ipwhoLookup(ip);
      if (enriched != null) {
        _log.info('ipwho enrichment: isp=${enriched['isp']}');
        return ConnectionInfo(
          ip: ip,
          isp: enriched['isp'] ?? 'Unknown',
          city: enriched['city'] ?? '',
          region: enriched['region'] ?? 'Unknown',
          country: country.isNotEmpty ? country : (enriched['country_code'] ?? ''),
          server: colo.isEmpty ? 'Cloudflare' : 'Cloudflare $colo',
        );
      }
      return ConnectionInfo(
        ip: ip,
        country: country,
        server: colo.isEmpty ? 'Cloudflare' : 'Cloudflare $colo',
      );
    }

    try {
      final resp = await http.get(Uri.parse(_ipifyUrl)).timeout(_shortTimeout);
      if (resp.statusCode == 200) {
        final data = jsonDecode(resp.body) as Map<String, dynamic>;
        return ConnectionInfo(
          ip: (data['ip'] as String?) ?? 'Unknown',
          server: 'Cloudflare',
        );
      }
    } catch (_) {}

    return const ConnectionInfo(server: 'Cloudflare');
  }

  @override
  Future<LatencyResult> testLatency({
    int samples = 10,
    ProgressCallback? onProgress,
  }) async {
    _cancelled = false;
    final latencyUrl = Uri.parse(_downUrl.replaceFirst('{}', '0'));
    final times = <double>[];

    for (var i = 0; i < samples; i++) {
      if (_cancelled) break;
      try {
        final sw = Stopwatch()..start();
        final resp = await http
            .get(latencyUrl, headers: _browserHeaders)
            .timeout(_shortTimeout);
        // Read body to EOF — http.get already does this, but reference
        // it to ensure the compiler doesn't elide.
        resp.bodyBytes.length;
        sw.stop();
        final elapsedMs = sw.elapsedMicroseconds / 1000.0;
        times.add(elapsedMs);
        onProgress?.call(LatencySample(
          current: i + 1,
          total: samples,
          valueMs: elapsedMs,
        ));
      } catch (_) {
        // Individual sample failed — keep going, fromSamples handles
        // "all failed" → failed=true.
      }
    }
    return LatencyResult.fromSamples(times);
  }

  @override
  Future<SpeedResult> testDownload({ProgressCallback? onProgress}) async {
    _cancelled = false;
    _client = http.Client();
    final results = <double>[];
    try {
      for (var i = 0; i < _downSizes.length; i++) {
        if (_cancelled) break;
        final (size, label) = _downSizes[i];
        try {
          final req = http.Request(
            'GET',
            Uri.parse(_downUrl.replaceFirst('{}', size.toString())),
          );
          req.headers.addAll(_browserHeaders);
          final sw = Stopwatch()..start();
          final resp = await _client!.send(req).timeout(_chunkTimeout);
          var totalBytes = 0;
          var nextReport = _progressEvery;
          await for (final piece in resp.stream) {
            if (_cancelled) break;
            totalBytes += piece.length;
            if (totalBytes >= nextReport) {
              final elapsedSec = sw.elapsedMicroseconds / 1000000.0;
              if (elapsedSec > 0) {
                onProgress?.call(DownloadProgress(
                  label: label,
                  speedMbps: (totalBytes * 8) / (elapsedSec * 1000000),
                  bytes: totalBytes,
                  current: i + 1,
                  total: _downSizes.length,
                ));
              }
              nextReport += _progressEvery;
            }
          }
          sw.stop();
          final elapsedSec = sw.elapsedMicroseconds / 1000000.0;
          if (elapsedSec <= 0) continue;
          final speed = (totalBytes * 8) / (elapsedSec * 1000000);
          results.add(speed);
          _log.info(
              'Download chunk $label: ${speed.toStringAsFixed(2)} Mbps in ${elapsedSec.toStringAsFixed(2)}s');
          onProgress?.call(DownloadChunk(
            label: label,
            speedMbps: speed,
            current: i + 1,
            total: _downSizes.length,
          ));
        } catch (e) {
          _log.warning('Download chunk $label failed: $e');
          onProgress?.call(ChunkError(
            label: label,
            error: e.toString(),
            isDownload: true,
          ));
        }
      }
    } finally {
      _client?.close();
      _client = null;
    }
    if (results.isEmpty) {
      throw const BackendError('all download attempts failed');
    }
    return SpeedResult.topHalf(results);
  }

  @override
  Future<SpeedResult> testUpload({ProgressCallback? onProgress}) async {
    _cancelled = false;
    _client = http.Client();
    final results = <double>[];
    final rng = Random();
    try {
      for (var i = 0; i < _upSizes.length; i++) {
        if (_cancelled) break;
        final (size, label) = _upSizes[i];
        try {
          final body = Uint8List(size);
          for (var j = 0; j < size; j++) {
            body[j] = rng.nextInt(256);
          }
          final sw = Stopwatch()..start();
          final resp = await _client!.post(
            Uri.parse(_upUrl),
            headers: {
              ..._browserHeaders,
              'Content-Type': 'application/octet-stream',
            },
            body: body,
          ).timeout(_chunkTimeout);
          resp.bodyBytes.length; // drain
          sw.stop();
          final elapsedSec = sw.elapsedMicroseconds / 1000000.0;
          if (elapsedSec <= 0) continue;
          final speed = (size * 8) / (elapsedSec * 1000000);
          results.add(speed);
          _log.info(
              'Upload chunk $label: ${speed.toStringAsFixed(2)} Mbps in ${elapsedSec.toStringAsFixed(2)}s');
          onProgress?.call(UploadChunk(
            label: label,
            speedMbps: speed,
            current: i + 1,
            total: _upSizes.length,
          ));
        } catch (e) {
          _log.warning('Upload chunk $label failed: $e');
          onProgress?.call(ChunkError(
            label: label,
            error: e.toString(),
            isDownload: false,
          ));
        }
      }
    } finally {
      _client?.close();
      _client = null;
    }
    if (results.isEmpty) {
      throw const BackendError('all upload attempts failed');
    }
    return SpeedResult.topHalf(results);
  }

  @override
  void cancel() {
    _cancelled = true;
    _client?.close();
  }
}
